doc/optimization.md

```diff
@@ -0,0 +1,86 @@
+# RaptorJIT optimization guide
+
+The first step to optimize an application is to characterize its most
+significant performance problem. First profile the application with
+`vmprofile` and then interpret the results using this step-by-step
+method:
+
+```lua
+-- RaptorJIT optimization procedure expressed as pseudo-code.
+function optimize ()
+   -- Spending nontrivial time in the interpreter? That's a bug - stop it!
+   if vmprofile.interpreted > 1% then
+      eliminate_interpreter()
+   -- Spend time in the garbage collector? That's a bug too - stop it!
+   elseif vmprofile.garbagecollection > 1% then
+      eliminate_allocation()
+   -- If you are not stuck in interpreter or GC then take a deeper look...
+   else
+      -- Spend time in JIT code? Check how efficient it is.
+      if vmprofile.head + vmprofile.loop > moderate then 
+         optimize_jit()
+      end
+      -- Spend time in the FFI? Check how efficient that is.
+      if vmprofile.ffi > moderate then
+          optimize_ffi()
+      end
+   end
+end
+```
+
+The control flow is significant in this method. For example, if the
+application is spending time in the interpreter then _that_ is the
+problem. You would need to eliminate this interpreter use before it
+makes sense look into other issues like garbage collection (which is
+often a secondary effect of running interpreted.)
+
+## Eliminate interpreter mode
+
+If your application is spending significant time in the interpreter
+then you need to look for _blacklisted_ code.
+
+Code is blacklisted when repeated attempts at JIT compilation all fail.
```

Comment:
This description seems to be oversimplified.

I had assumed that the first step is to find *blacklisted* code and then the second step is to understand why using *aborts*. However, in practice it seems like code that fails to JIT will not necessarily be blacklisted.

Here is a simple example with LuaJIT v2.1:

```
luajit/src$ ./luajit -jv <<EOF
for i = 1, 100e6 do 
   f = function() end
end
EOF
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
[TRACE --- (command line):1 -- NYI: bytecode 51]
```

This code repeatedly fails to JIT because it creates a closure (NYI `FNEW` bytecode) but it is never reported to be blacklisted (not by `-jdump` either.)

So this makes things a bit more complicated. You can't simply look for blacklisting messages and work backwards from those because they will not necessarily appear. So which messages _should_ you care about? Could be that you need to find source locations where traces have repeatedly been started and aborted, that were never followed by a successful trace. Could be that tooling can aid this.

@javierguerragiraldez Do you happen to know what makes some traces be blacklisted and others not? When is a bytecode blacklisted, exactly, if not simply because it repeatedly aborts when used as the starting point of a trace?